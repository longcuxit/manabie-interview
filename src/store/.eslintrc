{
  "extends": "../../.eslintrc",
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": [
          "components/**",
          "pages",
          "pages/**",
          "!./**"
        ]
      }
    ]
  }
}