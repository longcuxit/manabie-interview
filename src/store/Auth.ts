import { AuthModel, UserModel, UserStatus } from "models/User";
import Service from "service";
import { createLocalStorage } from "utils";
import { Store } from "utils/Store";
import { AuthActions, AuthState } from "./Auth.type";

const questUser: UserModel = {
  username: "Quest",
  id: "Quest",
  status: UserStatus.ACTIVE,
  created_date: "",
};

const authStorage = createLocalStorage<AuthModel | null>("AUTH", null);

const initialState: AuthState = { auth: authStorage.get(), user: questUser };

const store = new Store<AuthState, AuthActions>(
  initialState,
  ({ partial }) => ({
    async login(username: string, password: string): Promise<void> {
      const auth = await Service.login(username, password);
      authStorage.set(auth);
      partial({ auth });
    },

    async logout(): Promise<void> {
      authStorage.set(null);
      partial({ auth: null, user: questUser });
    },
  })
);

export const useAuth = store.createHook();

export const useAuthAction = store.createHookAction();

export * from "./Auth.type";
