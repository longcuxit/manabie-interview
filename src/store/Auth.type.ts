import { AuthModel, UserModel } from "models/User";

export interface AuthState {
  auth: AuthModel | null;
  user: UserModel;
}

export interface AuthActions {
  login(username: string, password: string): Promise<void>;
  logout(): Promise<void>;
}
