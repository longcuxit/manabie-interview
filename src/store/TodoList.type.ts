import { TodoModel } from "models/Todo";

export interface TodoState {
  todos: Map<string, TodoModel>;
  countActive: number;
}

export interface TodoActions {
  initialize(): Promise<void>;
  create(content: string): Promise<void>;
  remove(id: string): Promise<void>;
  updateStatus(id: string, active: boolean): Promise<void>;
  updateContent(id: string, content: string): Promise<void>;
  clearAll(): Promise<void>;
  toggleAll(active: boolean): Promise<void>;
}
