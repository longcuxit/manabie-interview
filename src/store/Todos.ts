import { TodoModel, TodoStatus } from "models/Todo";
import Service from "service";
import { Store } from "utils/Store";

export interface TodoState {
  todos: Map<string, TodoModel>;
  countActive: number;
}

const initialState: TodoState = {
  todos: new Map<string, TodoModel>(),
  countActive: 0,
};

const store = new Store(initialState, ({ set, get }) => ({
  async initialize() {
    const todos = await Service.getTodos();
    const map = new Map<string, TodoModel>();
    let countActive = 0;

    todos.forEach((todo) => {
      if (todo.status === TodoStatus.ACTIVE) countActive++;
      map.set(todo.id, todo);
    });

    set({ todos: map, countActive });
  },

  async create(content: string) {
    const newTodo = await Service.createTodo(content);
    set(({ todos, countActive }) => {
      todos.set(newTodo.id, newTodo);

      return {
        todos: new Map(todos),
        countActive: countActive + +(newTodo.status === TodoStatus.ACTIVE),
      };
    });
  },

  async remove(id: string) {
    if (!(await Service.removeTodo(id))) return;

    set((state) => {
      const exist = state.todos.get(id)!;

      state.todos.delete(id);
      state.countActive -= +(exist.status === TodoStatus.ACTIVE);

      return { ...state, todos: new Map(state.todos) };
    });
  },

  async updateStatus(id: string, active: boolean) {
    const toStatus = active ? TodoStatus.ACTIVE : TodoStatus.COMPLETED;

    const todo = get().todos.get(id)!;
    if (todo.status === toStatus) return;

    if (!(await Service.updateTodoStatus(id, toStatus))) return;

    set((state) => {
      todo.status = toStatus;
      state.todos.set(id, { ...todo });
      state.countActive += +active * 2 - 1;

      return { ...state, todos: new Map(state.todos) };
    });
  },

  async updateContent(id: string, content: string) {
    const todo = get().todos.get(id)!;
    if (todo.content === content) return;
    if (!(await Service.updateTodoContent(id, content))) return;

    set((state) => {
      todo.content = content;
      state.todos.set(id, { ...todo });
      return { ...state, todos: new Map(state.todos) };
    });
  },

  async clearAll() {
    const { todos } = get();

    const deleteIds: string[] = [];
    get().todos.forEach((todo) => {
      if (todo.status === TodoStatus.COMPLETED) {
        deleteIds.push(todo.id);
      }
    });

    if (!(await Service.clearTodos(deleteIds))) return;
    deleteIds.forEach(todos.delete.bind(todos));

    set({ countActive: todos.size, todos: new Map(todos) });
  },

  async toggleAll(active: boolean) {
    const toStatus = active ? TodoStatus.ACTIVE : TodoStatus.COMPLETED;
    if (!(await Service.updateAllTodoStatus(toStatus))) return;

    set((state) => {
      state.todos.forEach((todo) => {
        state.todos.set(todo.id, { ...todo, status: toStatus });
      });

      state.countActive = active ? state.todos.size : 0;

      return { ...state, todos: new Map(state.todos) };
    });
  },
}));

export const useTodos = store.createHook();

export const useTodoActions = store.createHookAction();
