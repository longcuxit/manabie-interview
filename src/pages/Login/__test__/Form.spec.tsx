import { fireEvent, render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { useAuthAction } from "store/Auth";
import FormLogin from "../Form";

jest.mock("store/Auth", () => ({
  useAuthAction: jest.fn(),
}));

const mockUseAuthAction = useAuthAction as jest.Mock;

describe("pages/Login/Form:", () => {
  test("match snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <FormLogin />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test("submit form", () => {
    const login = jest.fn();
    mockUseAuthAction.mockImplementation(() => ({ login }));

    const { container } = render(
      <MemoryRouter>
        <FormLogin />
      </MemoryRouter>
    );
    fireEvent.click(container.querySelector("[type=submit]")!);

    expect(login).toBeCalledTimes(1);
  });
});
