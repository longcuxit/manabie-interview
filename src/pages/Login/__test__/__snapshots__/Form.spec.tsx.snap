// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Login/Form: match snapshot 1`] = `
<div>
  <form
    class=""
  >
    <label
      class="form-group mb-4 d-block"
    >
      <input
        class="form-control form-control-lg"
        placeholder="Enter a valid email address"
        type="email"
      />
      <span
        class="form-label"
      >
        Email address
      </span>
    </label>
    <label
      class="form-group mb-4 d-block"
    >
      <input
        class="form-control form-control-lg"
        placeholder="Enter password"
        type="password"
      />
      <span
        class="form-label"
      >
        Password
      </span>
    </label>
    <div
      class="d-flex justify-content-between align-items-center"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
        <label
          class="form-check-label"
          title=""
        >
          Remember me
        </label>
      </div>
      <a
        class="text-body"
        href="#!"
      >
        Forgot password?
      </a>
    </div>
    <div
      class="text-center text-lg-start mt-4 pt-2"
    >
      <button
        class="px-2 w-100 btn btn-primary btn-lg"
        type="submit"
      >
        Login
      </button>
      <p
        class="small fw-bold my-2 pt-1"
      >
        Don't have an account?
         
        <a
          class="link-danger"
          href="/#!"
        >
          Register
        </a>
      </p>
    </div>
  </form>
</div>
`;
